cross_layer_equalization

edmundw314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40521 cross_layer_equalization**

Differential Revision: [D22603548](https://our.internmc.facebook.com/intern/diff/D22603548)

edmundw314 committed:
cross_layer_equalization

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,89 @@
+import torch
+import torch.nn as nn
+
+def channel_range(tensor, axis =0):
```
NIT: Please dont use tensor as a variable name, torch.Tensor is a class in pytorch and the naming can be confusing

raghuramank100 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,89 @@
+    size_of_tensor_dim = len(tensor.size())
+    mins = tensor
+    maxs = tensor
+    for i in range(size_of_tensor_dim - 1, axis, -1):
```
Is the first loop trying to initialize the tensor?

edmundw314 replied:
What i'm trying to do is apply the torch.min/max on every dim except the dim specified by axis, didn't see any function already made for this, but the idea is given a n-dim tensor with size [a,b,c.... x], i wanted to apply the min function on all the dim after axis (first for loop) and then apply the min function on all the dim before axis (the second for loop)

raghuramank100 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,89 @@
+
+    # producing scaling factors to applied
+    zero_input = torch.zeros(output_channel_tensor1.size())
+    scaling_factors = torch.sqrt(torch.addcdiv(zero_input, output_channel_tensor1, input_channel_tensor2))
```
You could just do: 
``` python
scaling_factors = torch.sqrt(output_channel_tensor1/input_channel_tensor2)
```

raghuramank100 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,89 @@
+    input_channel_tensor2 = channel_range(tensor2, input_axis)
+
+    # producing scaling factors to applied
+    zero_input = torch.zeros(output_channel_tensor1.size())
```
Lets make sure that we add a small constant to input_channel_tensor2 so that we do not divide by zero.

raghuramank100 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,89 @@
+    scaling_factors = torch.unsqueeze(scaling_factors, output_axis)
+    r_scaling_factors = torch.unsqueeze(r_scaling_factors, input_axis)
+
+    for i in range(len(tensor1.size()) - 2):
```
Can you explain the purpose of the loop here?

edmundw314 replied:
this loop is for when dim of a tensor is greater than 2, when dealing with a conv2d we get a 4d tensor and I was trying to pad the scaling_factors tensors with the extra dims so i could then multiply in the next few lines (broadcasting didn't work for me when I tried the first time, can give it another go)

jerryzh168 replied:
original scaling factor is one number, I think it might be clearer to do something like this:
1. get a list of 1s with the same dimension as original tensor, e.g. sizes = [1] * tensor1.dim()
2. set particular axis to scaling factor,  sizes[out_axis] = scaling_factor

raghuramank100 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,89 @@
+    for i in range(len(tensor2.size()) - 2):
+        scaling_factors = torch.unsqueeze(scaling_factors, -1)
+
+    mod_tensor1 = torch.mul(tensor1, r_scaling_factors)
```
You could also do this simply with * instead of mul.

raghuramank100 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,89 @@
+    else:
+        raise TypeError("Only Linear and Conv2d modules are supported at this time")
+
+    tensor1, tensor2 = scaling_channels(module1.weight, module2.weight,
```
Would it be better to modify the tensors in place, so that you dont need to return them and then assign them.

raghuramank100 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,89 @@
+    tensor1, tensor2 = scaling_channels(module1.weight, module2.weight,
+                module1_output_channel_axis, module2_input_channel_axis)
+
+    module1.weight = nn.parameter.Parameter(tensor1)
```
Do we need to convert tensor1 to a parameter prior to assignment?

edmundw314 replied:
Will see if there is a work around, but was getting an error at one point that the module1 didn't like the fact that I was trying to change module1.weight to something that wasn't a parameter, double checking if this error is persisting

jerryzh168 replied:
you can try setting model to eval mode before equalization: `model.eval()`

raghuramank100 commented on test/quantization/test_equalize.py:
```diff
@@ -0,0 +1,36 @@
+        output_axis = 0
+        input_axis = 1
+
+        mod_tensor1, mod_tensor2 = _equalize.scaling_channels(tensor1, tensor1, output_axis, input_axis)
```
This should be tensor1, tensor2

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,115 @@
+    return mod_tensor1, mod_tensor2
+
+
+def cross_layer_equalization(module1, module2):
```
equalization is only applied on weights, I think it's better to have this function only accept weights, you can do verification outside of this function.

Also looks like for both conv and linear the channel axis for first module is always 0 and always 1 for second module, so we don't really need to check that. Although you can do some validation checking the dimension of these axis match.

jerryzh168 replied:
nit: we want to use verbs for functions that does something, so: cross_layer_equalization --> `cross_layer_equalize` or just `_equalize_impl`

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_equalize.py:
```diff
@@ -0,0 +1,100 @@
+import copy
+
+class TestEqualizeEager(QuantizationTestCase):
+    def channels_equalized_test(self, tensor1, tensor2, output_axis, input_axis):
```
nit: naming for test should be `test_xxx`

jerryzh168 replied:
naming for helper function is camelcase I think, e.g. checkChannelsEqualized

edmundw314 replied:
im using this function as a helper function for the other defined tests, if you name it test_xxx wouldn't the unittester think its a test function?

jerryzh168 replied:
please see my second comment

jerryzh168 commented on test/quantization/test_equalize.py:
```diff
@@ -0,0 +1,100 @@
+        # tensor2's output
+        self.assertEqual(output_channel_tensor1, input_channel_tensor2)
+
+    def get_module(self, model, name):
```
same, this should be getModule

edmundw314 replied:
also a helper function, do we still name it as test_xxx?

jerryzh168 replied:
no, it should be getModule if it's a helper function

jerryzh168 commented on test/quantization/test_equalize.py:
```diff
@@ -0,0 +1,100 @@
+
+        self.channels_equalized_test(mod_tensor1, mod_tensor2, module1_output_channel_axis, module2_input_channel_axis)
+
+    def test_convergence_test(self):
```
naming is a bit confusing, could you change this

jerryzh168 commented on test/quantization/test_equalize.py:
```diff
@@ -0,0 +1,100 @@
+        self.assertFalse(_equalize.converged_test(dictionary_1, dictionary_2, 1e-6))
+
+    def test_equalize(self):
+        class chain_module(nn.Module):
```
module is Captilized typically, should be ChainModule

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+
+    scaling_channels(module1, module2, module1_output_channel_axis, module2_input_channel_axis)
+
+def equalize(model, paired_modules_list, threshold):
```
would probably make sense to give a default value for `threshold`

jerryzh168 replied:
also, you can provide a inplace option, when inplace is false you can copy the model, and all the following modification can be inplace. example: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize.py#L209

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+    '''
+    name_to_module = {}
+    copies = {}
+    name_set = {name for pair in paired_modules_list for name in pair}
```
is this to dedup? you can just do name_set = set(paired_modules_list) I think?

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+
+            cross_layer_equalization(name_to_module[pair[0]], name_to_module[pair[1]])
+
+def converged_test(curr_modules, prev_modules, threshold):
```
maybe `converged_test` --> `converged`

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+    for name, module in model.named_modules():
+        if name in name_set:
+            name_to_module[name] = module
+            copies[name] = None
```
copies sounds vague, would be good to rename it to something like `previous_name_to_module`

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+    assert(mins.size()[0] == input.size()[axis])
+    return maxs - mins
+
+def scaling_channels(module1, module2, output_axis=0, input_axis=1):
```
I think you can remove this function and move the implementation to cross_layer_equalization function

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+
+    # formatting the scaling (1D) tensors to be applied on the given argument tensors
+    # pads axis to (1D) tensors to then be broadcasted
+    size1 = [1 for x in range(len(tensor1.size()))]
```
can be [1] *tensor1.dim()

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+    # formatting the scaling (1D) tensors to be applied on the given argument tensors
+    # pads axis to (1D) tensors to then be broadcasted
+    size1 = [1 for x in range(len(tensor1.size()))]
+    size1[output_axis] = tensor1.size()[output_axis]
```
tensor1.size()[output_axis] --> tensor1.dim(output_axis)

edmundw314 replied:
the dim() function for tensors just returns the number of axises, it doesn't query the axis you want like tensor1.size()[output_axis]

jerryzh168 replied:
Oh I think I meant, tensor1.size(output_axis) here

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+    tensor1 = tensor1 * r_scaling_factors
+    tensor2 = tensor2 * scaling_factors
+
+    module1.weight.data = tensor1
```
I'm not sure if assigning data is a good idea actually

edmundw314 replied:
the previous version was module1.weight = nn.parameter.Parameter(tensor1), would that be better?

jerryzh168 replied:
yeah I think we should be using that, we also do that in the conv bn fusion: https://github.com/pytorch/pytorch/blob/ad30d465d5dfe10e0e5ae9d48813a1d60a71481e/torch/nn/utils/fusion.py#L25

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+    ranges of the second tensors' input channel
+
+    '''
+    tensor1 = module1.weight
+    tensor2 = module2.weight
```
tensor1, tensor2 should probably be weight1 weight2

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,133 @@
+    tensor2 = module2.weight
+    bias = module1.bias
+
+    output_channel_tensor1 = channel_range(tensor1, output_axis)
```
weight1_range?

jerryzh168 reviewed (changes requested):
comments are inline, in general please make naming more intuitive

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,111 @@
+    module1.bias.data = bias
+    module2.weight.data = weight2
+
+def equalize(model, paired_modules_list, threshold = 1e-4):
```
please fix lint errors

jerryzh168 commented on test/quantization/test_equalize.py:
```diff
@@ -0,0 +1,86 @@
+    def test_equalize(self):
+        class ChainModule(nn.Module):
+            def __init__(self):
+                """
+                In the constructor we instantiate two nn.Linear modules and assign them as
+                member variables.
+                """
```
these comments are probably not necessary, would be good to add a comment for the test describing what we are testing though.

jerryzh168 commented on test/quantization/test_equalize.py:
```diff
@@ -0,0 +1,86 @@
+                self.linear3 = nn.Linear(5, 6)
+
+            def forward(self, x):
+                """
+                In the forward function we accept a Tensor of input data and we must return
+                a Tensor of output data. We can use Modules defined in the constructor as
+                well as arbitrary operators on Tensors.
+                """
```
these are not necessary as well

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,111 @@
+    ranges of the second tensors' input channel
+
+    '''
+    if module1.weight.size()[output_axis] != module2.weight.size()[input_axis]:
```
you still need to check the types for the modules, but the output_axis/input_axis can be fixed.

jerryzh168 replied:
like:
```
_supported_types = [torch.nn.Conv2d, torch.nn.Linear]
if type(module1) not in _supported_types or type(module2) not in _supported_types:
    raise ValueError("module type not supported:", type(module1), " ",type(module2))
```
you can do this here or in `equalize` function.

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,120 @@
+
+    summed_norms = 0
+    for name in curr_modules.keys():
+        if prev_modules[name] is None:
+            return False
```
this should be outside of the loop

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,120 @@
+    maxs = input
+
+    # reshape input to make specified axis the last axis in input
+    temp = list(range(size_of_tensor_dim))
+    temp[size_of_tensor_dim - 1] = axis
+    temp[axis] = size_of_tensor_dim - 1
+
+    mins = mins.permute(temp)
+    maxs = maxs.permute(temp)
+
+    # minimizing over all axises except the last axis
+    for i in range(size_of_tensor_dim - 1):
+        mins = torch.min(mins, 0)[0]
+        maxs = torch.max(maxs, 0)[0]
```
maybe you can do something like this to make it more clear: 
1. figure out the axis you need to reduce, by getting all the axis and remove the input axis:
```
axis = list(range(input.dim())
axis.remove(axis)
```
2. min/max over the reduce axis
```
def multi_max(input, axes, keepdim=False):
    '''
    Performs `torch.max` over multiple dimensions of `input`
    '''
    axes = sorted(axes)
    maxed = input
    for axis in reversed(axes):
        maxed, _ = maxed.max(axis, keepdim)
    return maxed
```

jerryzh168 reviewed (approved):
please address some final comments before landing

jerryzh168 commented:
Also please fix the CI errors

vkuzo reviewed (commented):
Would it be possible to add the context to both the PR and the diff summary (goal of this, any reference papers, etc)?

supriyar commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,120 @@
+
+def channel_range(input, axis=0):
+    ''' finds the range of weights associated with a specific channel
+
```
Extra line here can be removed

supriyar commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,120 @@
+    scaling_factors = torch.sqrt(weight1_range / weight2_range)
+    r_scaling_factors = torch.reciprocal(scaling_factors)
+
+    bias = bias * r_scaling_factors
```
Maybe calling them inverse_scaling_factors as opposed to r_scaling_factors for better readability

supriyar commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,120 @@
+
+    '''
+    if len(curr_modules) != len(prev_modules):
+        raise TypeError("in compatiables modules in convergence condition")
```
in compatibiles -> incompatible

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

z-a-f commented on test/quantization/test_equalize.py:
```diff
@@ -0,0 +1,95 @@
+        # tensor2's output
+        self.assertEqual(output_channel_tensor1, input_channel_tensor2)
+
+    def getModule(self, model, name):
+        ''' Given the name is a submodule to a model, return the submodule
+        '''
+        curr = model
+        name = name.split('.')
+        for subname in name:
+            curr = curr._modules[subname]
+        return curr
```
nit: I think we can put this into external utility file. But this doesn't have to be part of this PR

z-a-f commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,130 @@
+    ''' Applies 'torch.max' over the given axises
+    '''
+    axis_list.sort(reverse=True)
+    reduced_input = input
```
There is no need for this line, as this is not a copy, but just getting the reference. The min/max function is not inplace, so you should not worry about input being modified

z-a-f commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,130 @@
+    ''' finds the range of weights associated with a specific channel
+    '''
+    size_of_tensor_dim = len(input.size())
+    mins = input
+    maxs = input
```
Same here -- unless you are using it in a loop, and need to initialize the variables, this doesn't really do much

z-a-f commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,130 @@
+    mins = min_over_ndim(mins, axis_list)
+    maxs = max_over_ndim(maxs, axis_list)
+
+    assert(mins.size()[0] == input.size()[axis])
```
Can you, please, add a comment here telling the user why the error happened? Also, note that the parenthesis in assertions are dangerous :) Sauce: https://stackoverflow.com/a/3112178

```python
    assert mins.size()[0] == input.size()[axis], "Some useful error message here..."
```

z-a-f commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,130 @@
+    ranges of the second tensors' input channel
+
+    '''
+    _supported_types = [torch.nn.Conv2d, torch.nn.Linear]
```
We should probably move this declaration to the top of the file. That way the developers will know exactly what is supported without a need to search for it.

z-a-f commented on torch/quantization/_equalize.py:
```diff
@@ -0,0 +1,130 @@
+
+    '''
+    if len(curr_modules) != len(prev_modules):
+        raise TypeError("incompatiable modules in convergence condition")
```
nit: Maybe say that the sizes are not compatible? In case the user is confused what exactly is not compatible.

z-a-f reviewed (commented):
Love it!

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

edmundw314 committed:
Update on "cross_layer_equalization"


Differential Revision: [D22214788](https://our.internmc.facebook.com/intern/diff/D22214788)

[ghstack-poisoned]

jerryzh168 commented:
is this deprecated? if so please close the commit
